packages/back-end/src/models/SegmentModel.ts

```diff
@@ -1,108 +1,76 @@
-  if (usingFileConfigForSegments()) {
-    return getConfigSegments(organization).filter((s) => s.id === id)[0];
+export class SegmentModel extends BaseClass {
+  protected canRead(): boolean {
+    return this.context.permissions.canReadSingleProjectResource("");
   }
-
-  const doc = await SegmentModel.findOne({ id, organization });
-
-  return doc ? toInterface(doc) : null;
-}
-
-export async function findSegmentsByOrganization(organization: string) {
-  // If using config.yml & the org doesn't have the env variable STORE_SEGMENTS_IN_MONGO,
-  // immediately return the list from there
-  if (usingFileConfigForSegments()) {
-    return getConfigSegments(organization);
+  protected canCreate(): boolean {
+    return this.context.permissions.canCreateSegment();
   }
-
-  return (await SegmentModel.find({ organization })).map(toInterface);
-}
-
-export async function findSegmentsByDataSource(
-  datasource: string,
-  organization: string
-) {
-  // If using config.yml & the org doesn't have the env variable STORE_SEGMENTS_IN_MONGO,
-  // immediately return the list from there
-  if (usingFileConfigForSegments()) {
-    return getConfigSegments(organization).filter(
-      (s) => s.datasource === datasource
-    );
+  protected canUpdate(): boolean {
+    return this.context.permissions.canUpdateSegment();
   }
+  protected canDelete(): boolean {
+    return this.context.permissions.canDeleteSegment();
+  }
+  protected useConfigFile(): boolean {
+    if (usingFileConfig() && !STORE_SEGMENTS_IN_MONGO) {
+      return true;
+    }
+    return false;
+  }
+  protected getConfigDocuments() {
+    if (!this.useConfigFile) return [];
 
-  return (await SegmentModel.find({ datasource, organization })).map(
-    toInterface
-  );
-}
-
-export async function deleteSegmentById(id: string, organization: string) {
-  // If using config.yml & the org doesn't have the env variable STORE_SEGMENTS_IN_MONGO,
-  // immediately throw error
-  if (usingFileConfigForSegments()) {
-    throw new Error("Cannot delete. Segments are being managed by config.yml");
+    return getConfigSegments(this.context.org.id);
   }
+  public async getByDataSource(
+    datasourceId: string
+  ): Promise<SegmentInterface[]> {
+    const allSegments = await this.getAll();
 
-  await SegmentModel.deleteOne({ id, organization });
-}
+    return allSegments.filter((segment) => segment.datasource === datasourceId);
+  }
 
-export async function updateSegment(
-  id: string,
-  organization: string,
-  updates: Partial<SegmentInterface>
-) {
-  // If using config.yml & the org doesn't have the env variable STORE_SEGMENTS_IN_MONGO,
-  // immediately return the list from there
-  if (usingFileConfigForSegments()) {
-    throw new Error("Cannot update. Segments are being managed by config.yml");
+  protected async beforeCreate() {
+    if (this.useConfigFile()) {
+      throw new Error(
+        "Cannot create. Segments are being managed by config.yml"
+      );
+    }
```

Comment:
This logic of blocking writes when using config should probably live in BaseModel instead.